### ISSUE TITLE
AOT compilation fails on Windows after the update to mono 2018-06

### ISSUE THREAD

**Issue by pjcollins** [xamarin/xamarin-android#2290]
https://github.com/xamarin/xamarin-android/commit/606675b59f52595e3030c529de4c856fb347edd8 appears to have broken AOT compilation in certain cases.

### Steps to Reproduce

1. msbuild TodoAzure.Droid.csproj /t:SignAndroidPackage /p:Configuration=Release /p:AotAssemblies=true /p:EnableLlvm=true
https://github.com/xamarin/xamarin-forms-samples/tree/master/WebServices/TodoAzurePush

### Actual Behavior
```
[aot-compiler stdout] Mono Ahead of Time compiler - compiling assembly C:\agent\_work\r8\a\xamarin-forms-samples\WebServices\TodoAzurePush\Droid\obj\Release\81\android\assets\PCLCrypto.dll (TaskId:288)
  [aot-compiler stdout] AOTID 5E31DBA6-D7C4-378F-663E-D6E8726A1DF9 (TaskId:288)
  [aot-compiler stdout] Executing opt: "C:\PROGRA~2\MICROS~1\2017\ENTERP~1\MSBuild\Xamarin\Android\opt" -f -O2 -disable-tail-calls -o "C:\agent\_work\r8\a\XAMARI~2\WEBSER~1\TO3284~1\Droid\obj\Release\81\aot\ARMEAB~1\PCLCrypto.dll\temp.opt.bc" "C:\agent\_work\r8\a\XAMARI~2\WEBSER~1\TO3284~1\Droid\obj\Release\81\aot\ARMEAB~1\PCLCrypto.dll\temp.bc" (TaskId:288)
  [AOT] MONO_PATH="C:\agent\_work\r8\a\xamarin-forms-samples\WebServices\TodoAzurePush\Droid\obj\Release\81\android\assets" MONO_ENV_OPTIONS="" "C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\Xamarin\Android\cross-arm" --llvm --aot=outfile=C:\agent\_work\r8\a\XAMARI~2\WEBSER~1\TO3284~1\Droid\obj\Release\81\aot\ARMEAB~1\libaot-Xamarin.Android.Support.Animated.Vector.Drawable.dll.so,asmwriter,mtriple=armv7-linux-gnueabi,tool-prefix=C:\PROGRA~2\Android\ANDROI~2\NDK-BU~1\TOOLCH~1\ARM-LI~1.9\prebuilt\WINDOW~1\bin\arm-linux-androideabi-,ld-flags=C:\PROGRA~2\Android\ANDROI~2\NDK-BU~1\TOOLCH~1\ARM-LI~1.9\prebuilt\WINDOW~1\lib\gcc\ARM-LI~1\49696B~1.X\libgcc.a;C:\PROGRA~2\Android\ANDROI~2\NDK-BU~1\PLATFO~1\ANDROI~2\arch-arm\usr\lib\libc.so;C:\PROGRA~2\Android\ANDROI~2\NDK-BU~1\PLATFO~1\ANDROI~2\arch-arm\usr\lib\libm.so,llvm-path=C:\PROGRA~2\MICROS~1\2017\ENTERP~1\MSBuild\Xamarin\Android,temp-path=C:\agent\_work\r8\a\XAMARI~2\WEBSER~1\TO3284~1\Droid\obj\Release\81\aot\ARMEAB~1\Xamarin.Android.Support.Animated.Vector.Drawable.dll C:\agent\_work\r8\a\xamarin-forms-samples\WebServices\TodoAzurePush\Droid\obj\Release\81\android\assets\Xamarin.Android.Support.Animated.Vector.Drawable.dll (TaskId:288)
  [aot-compiler stdout] Executing llc: "C:\PROGRA~2\MICROS~1\2017\ENTERP~1\MSBuild\Xamarin\Android\llc"  -mattr=+vfp2,-neon,+d16 -asm-verbose=false -mtriple=armv7-linux-gnueabi -disable-gnu-eh-frame -enable-mono-eh-frame -mono-eh-frame-symbol=mono_aot_PCLCrypto_eh_frame -disable-tail-calls -relocation-model=pic -o "C:\agent\_work\r8\a\XAMARI~2\WEBSER~1\TO3284~1\Droid\obj\Release\81\aot\ARMEAB~1\PCLCrypto.dll\temp-llvm.s" "C:\agent\_work\r8\a\XAMARI~2\WEBSER~1\TO3284~1\Droid\obj\Release\81\aot\ARMEAB~1\PCLCrypto.dll\temp.opt.bc" (TaskId:288)
  [aot-compiler stderr] Wrote crash dump file "C:\Users\XAMARI~1\AppData\Local\Temp\llc.exe-b94cbe.dmp" (TaskId:288)
  [aot-compiler stderr] AOT of image C:\agent\_work\r8\a\xamarin-forms-samples\WebServices\TodoAzurePush\Droid\obj\Release\81\android\assets\PCLCrypto.dll failed. (TaskId:288)
C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(2725,3): error XA3001: Could not AOT the assembly: PCLCrypto.dll [C:\agent\_work\r8\a\xamarin-forms-samples\WebServices\TodoAzurePush\Droid\TodoAzure.Droid.csproj]

  ...

[AOT] MONO_PATH="C:\agent\_work\r8\a\xamarin-forms-samples\WebServices\TodoAzurePush\Droid\obj\Release\81\android\assets" MONO_ENV_OPTIONS="" "C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\Xamarin\Android\cross-arm" --llvm --aot=outfile=C:\agent\_work\r8\a\XAMARI~2\WEBSER~1\TO3284~1\Droid\obj\Release\81\aot\ARMEAB~1\libaot-Xamarin.Android.Support.Annotations.dll.so,asmwriter,mtriple=armv7-linux-gnueabi,tool-prefix=C:\PROGRA~2\Android\ANDROI~2\NDK-BU~1\TOOLCH~1\ARM-LI~1.9\prebuilt\WINDOW~1\bin\arm-linux-androideabi-,ld-flags=C:\PROGRA~2\Android\ANDROI~2\NDK-BU~1\TOOLCH~1\ARM-LI~1.9\prebuilt\WINDOW~1\lib\gcc\ARM-LI~1\49696B~1.X\libgcc.a;C:\PROGRA~2\Android\ANDROI~2\NDK-BU~1\PLATFO~1\ANDROI~2\arch-arm\usr\lib\libc.so;C:\PROGRA~2\Android\ANDROI~2\NDK-BU~1\PLATFO~1\ANDROI~2\arch-arm\usr\lib\libm.so,llvm-path=C:\PROGRA~2\MICROS~1\2017\ENTERP~1\MSBuild\Xamarin\Android,temp-path=C:\agent\_work\r8\a\XAMARI~2\WEBSER~1\TO3284~1\Droid\obj\Release\81\aot\ARMEAB~1\Xamarin.Android.Support.Annotations.dll C:\agent\_work\r8\a\xamarin-forms-samples\WebServices\TodoAzurePush\Droid\obj\Release\81\android\assets\Xamarin.Android.Support.Annotations.dll (TaskId:288)
C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(2725,3): error XA3001: Could not AOT the assembly: Xamarin.Android.Arch.Lifecycle.Common.dll [C:\agent\_work\r8\a\xamarin-forms-samples\WebServices\TodoAzurePush\Droid\TodoAzure.Droid.csproj]
C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(2725,3): error XA3001: Could not AOT the assembly: Xamarin.Android.Arch.Lifecycle.Runtime.dll [C:\agent\_work\r8\a\xamarin-forms-samples\WebServices\TodoAzurePush\Droid\TodoAzure.Droid.csproj]
C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(2725,3): error XA3001: Could not AOT the assembly: Xamarin.Android.Support.Animated.Vector.Drawable.dll [C:\agent\_work\r8\a\xamarin-forms-samples\WebServices\TodoAzurePush\Droid\TodoAzure.Droid.csproj]
C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(2725,3): error XA3001: Could not AOT the assembly: Xamarin.Android.Support.Annotations.dll [C:\agent\_work\r8\a\xamarin-forms-samples\WebServices\TodoAzurePush\Droid\TodoAzure.Droid.csproj]
C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(2725,3): error XA3001: Could not AOT the assembly: Microsoft.Azure.Mobile.Client.dll [C:\agent\_work\r8\a\xamarin-forms-samples\WebServices\TodoAzurePush\Droid\TodoAzure.Droid.csproj]
C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\Xamarin\Android\Xamarin.Android.Common.targets(2725,3): error XA3001: Could not AOT the assembly: Newtonsoft.Json.dll [C:\agent\_work\r8\a\xamarin-forms-samples\WebServices\TodoAzurePush\Droid\TodoAzure.Droid.csproj]
Done executing task "Aot" -- FAILED. (TaskId:288)
```
### Version Information
XA version: 9.1.99.130 (xamarin-android/HEAD/a9183e5) 

### Log File
http://xqa.blob.core.windows.net/gist/log-fda8f67a562b4fa39ee13f6de4e8c1d1.txt

> VS bug [#703854](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/703854), VS bug [#723070](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/723070)

**Comment by pjcollins** [comment]
Good to see other windows tests running as part of build validation, though this makes me wonder -- why did we merge this PR with at least one known regression? Do those tests report statuses to GH or are otherwise part of a merge checklist? Could they be?

@jonpryor What are the next steps in getting this resolved? Should a new issue be created in https://github.com/mono/mono?

**Comment by jonathanpeppers** [comment]
The Windows build won't succeed until after a PR bumping mono is merged, and a mono bundle uploaded...

You see an error such as:
```
Error : The Windows build depends on a cached mono bundle, that may not be available yet. Try going back a few commits (with git reset) to download a different bundle.
```

So any mono bump _has_ to get merged without a CI build on Windows.

This might be something we can get rid of eventually, since the Mono team is working on providing their own bundle we can download.

**Comment by pjcollins** [comment]
Here's some additional context to help diagnose, created with the command shown below - [llc.zip](https://github.com/xamarin/xamarin-android/files/2596253/llc.zip)
> msbuild App1.csproj /t:SignAndroidPackage /p:Configuration=Release /p:AotAssemblies=true /p:EnableLlvm=true

**Comment by lateralusX** [comment]
Thanks, will take a look.

**Comment by lateralusX** [comment]
I have investigated this on both 32-bit and 64-bit VS 2015 LLVM build as well as 64-bit x86_64-w64-mingw32 and i686-w64-mingw32 and it only seems to reproduce on 32-bit LLVM tools build using i686-w64-mingw32 version 5.3.1 and 5.4.0 and Androids LLVM build seems to be using mingw 5.4.0. Looking deeper into the crash dumps generated by the 32-bit mingw llc.exe (release optimized code without symbols so took some time to lay the puzzel) revealed that a call from :AddressingModeCombiner::findCommon to AddressingModeCombiner::MatchPhiSet incorrectly restores stack pointer, esp after the call that in turn will cause next load of a stack variable relative esp (optimized to not use frame pointer) to get incorrect value causing a crash on dereference. Since this is an compiler optimization bug it could be fixed in later gcc version (I have seen similar reports related to building LLVM under i686 gcc 6.2.0), but since the above versions seems to be current mingw versions distributed and used by MXE an option is to lower the optimization level of this function to make sure it doesn't trigger the optimization bug. I have opened https://github.com/mono/llvm/pull/19 in Mono's LLVM repository including a fix adjusting optimization for the problematic function. With that fix included in a local build llc.exe using i686-w64-mingw32 5.3.1 (using WSL) successfully compile the .bc file previously causing a crash.
